File: src/lib/tenantContract.js
import { z } from 'zod';
import siteFactoryManifest from '@/config/siteFactoryManifest.json';

const marketGeoSchema = z
  .object({
    region: z.string().optional(),
    placename: z.string().optional(),
    position: z.string().optional(),
    icbm: z.string().optional(),
  })
  .partial();

const marketSchema = z
  .object({
    marketName: z.string().optional(),
    primaryRegion: z.string().optional(),
    primaryMetro: z.string().optional(),
    heroKicker: z.string().optional(),
    heroHeadline: z.string().optional(),
    heroBody: z.string().optional(),
    ctaLabel: z.string().optional(),
    phoneDisplay: z.string().optional(),
    proofHeadline: z.string().optional(),
    geo: marketGeoSchema.optional(),
  })
  .partial();

const profileSchema = z.object({
  key: z.string().min(1),
  label: z.string().min(1),
  canonicalUrl: z.string().url(),
  primaryColor: z.string().optional(),
  accentColor: z.string().optional(),
  routeMode: z.string().optional(),
  enableChatWidget: z.boolean().optional(),
  market: marketSchema.optional(),
});

const tenantManifestSchema = z.object({
  profiles: z.array(profileSchema),
  hostnames: z.record(z.string(), z.string()),
});

const fallbackManifest = {
  profiles: [
    {
      key: 'jworden',
      label: 'J. Worden Asphalt Paving',
      canonicalUrl: 'https://www.jwordenasphaltpaving.com',
      routeMode: 'full-site',
      enableChatWidget: true,
    },
  ],
  hostnames: {
    'www.jwordenasphaltpaving.com': 'jworden',
    'jwordenasphaltpaving.com': 'jworden',
  },
};

const parsedManifest = tenantManifestSchema.safeParse(siteFactoryManifest);

if (!parsedManifest.success) {
  // Keep runtime resilient in production builds while surfacing contract drift in console.
  // The deploy gate script enforces strict validation in CI.
  console.warn('[tenant-contract] Manifest validation failed, using fallback manifest', parsedManifest.error?.issues || []);
}

const manifest = parsedManifest.success ? parsedManifest.data : fallbackManifest;

const profilesByKey = Object.fromEntries(manifest.profiles.map((p) => [p.key.toLowerCase(), p]));
const hostnames = Object.fromEntries(
  Object.entries(manifest.hostnames || {}).map(([host, tenant]) => [String(host).toLowerCase(), String(tenant).toLowerCase()]),
);

export const DEFAULT_TENANT = 'jworden';

export function getTenantByKey(key) {
  if (!key) return profilesByKey[DEFAULT_TENANT] || null;
  return profilesByKey[String(key).toLowerCase()] || null;
}

export function getTenantForHost(hostname) {
  if (!hostname) return getTenantByKey(DEFAULT_TENANT);
  const tenantKey = hostnames[String(hostname).toLowerCase()];
  return getTenantByKey(tenantKey || DEFAULT_TENANT);
}

export function getTenantForCurrentHost() {
  if (typeof window === 'undefined') return getTenantByKey(DEFAULT_TENANT);
  return getTenantForHost(window.location.hostname);
}

export function getAllTenantKeys() {
  return Object.keys(profilesByKey);
}

export function getHostTenantMap() {
  return { ...hostnames };
}

File: app/core/tenant_contract.py
"""tenant_contract.py — shared tenant manifest contract + loaders.

This module centralizes validation and reading of
src/config/siteFactoryManifest.json so backend services do not duplicate
tenant/hostname parsing logic.
"""

from __future__ import annotations

import json
import logging
from functools import lru_cache
from pathlib import Path

from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

_MANIFEST_PATH = (
    Path(__file__).resolve().parents[2] / "src" / "config" / "siteFactoryManifest.json"
)


class MarketGeo(BaseModel):
    region: str | None = None
    placename: str | None = None
    position: str | None = None
    icbm: str | None = None


class MarketConfig(BaseModel):
    marketName: str | None = None
    primaryRegion: str | None = None
    primaryMetro: str | None = None
    heroKicker: str | None = None
    heroHeadline: str | None = None
    heroBody: str | None = None
    ctaLabel: str | None = None
    phoneDisplay: str | None = None
    proofHeadline: str | None = None
    geo: MarketGeo | None = None


class TenantProfile(BaseModel):
    key: str = Field(min_length=1)
    label: str = Field(min_length=1)
    canonicalUrl: str = Field(min_length=1)
    primaryColor: str | None = None
    accentColor: str | None = None
    routeMode: str | None = None
    enableChatWidget: bool | None = None
    market: MarketConfig | None = None


class SiteFactoryManifest(BaseModel):
    profiles: list[TenantProfile] = Field(default_factory=list)
    hostnames: dict[str, str] = Field(default_factory=dict)


def infer_default_state(profile: TenantProfile) -> str:
    key = (profile.key or "").lower()
    if key == "jworden":
        return "VA"
    if key == "minnesota":
        return "MN"

    region = ((profile.market.primaryRegion if profile.market else "") or "").lower()
    if "minnesota" in region:
        return "MN"
    return "VA"


@lru_cache(maxsize=1)
def load_site_factory_manifest() -> SiteFactoryManifest:
    try:
        with _MANIFEST_PATH.open("r", encoding="utf-8") as f:
            raw = json.load(f)
        return SiteFactoryManifest.model_validate(raw)
    except Exception as exc:  # noqa: BLE001
        logger.warning(
            "Failed to load tenant manifest (%s). Falling back to jworden only.", exc
        )
        return SiteFactoryManifest(
            profiles=[
                TenantProfile(
                    key="jworden",
                    label="J. Worden Asphalt Paving",
                    canonicalUrl="https://www.jwordenasphaltpaving.com",
                    routeMode="full-site",
                    enableChatWidget=True,
                )
            ],
            hostnames={
                "www.jwordenasphaltpaving.com": "jworden",
                "jwordenasphaltpaving.com": "jworden",
            },
        )


def profiles_by_key() -> dict[str, TenantProfile]:
    manifest = load_site_factory_manifest()
    return {p.key: p for p in manifest.profiles}
